eland/ml/pytorch/external_models.py

```diff
@@ -0,0 +1,391 @@
+
+
+class DPREncoderWrapper(nn.Module):
+    """
+    AutoModel loading does not work for DPRContextEncoders, this only exists as
+    a workaround.
+    See: https://github.com/huggingface/transformers/issues/13670
+    """
+    _SUPPORTED_MODELS = {
+        transformers.DPRContextEncoder,
+        transformers.DPRQuestionEncoder,
+    }
+    _SUPPORTED_MODELS_STR = set([x.__name__ for x in _SUPPORTED_MODELS])
+
+    def __init__(self, model: Union[transformers.DPRContextEncoder, transformers.DPRQuestionEncoder]):
+        super(DPREncoderWrapper, self).__init__()
+        self._model = model
+
+    @staticmethod
+    def from_pretrained(model_id: str) -> Optional[Any]:
+
+        config = AutoConfig.from_pretrained(model_id)
+
+        def is_compatible() -> bool:
+            is_dpr_model = config.model_type == 'dpr'
+            has_architectures = len(config.architectures) == 1
+            is_supported_architecture = config.architectures[0] in DPREncoderWrapper._SUPPORTED_MODELS_STR
+            return is_dpr_model and has_architectures and is_supported_architecture
+
+        if is_compatible():
+            model = getattr(transformers, config.architectures[0]).from_pretrained(model_id, torchscript=True)
+            return DPREncoderWrapper(model)
+        else:
+            return None
+
+    def forward(self, input_ids: Tensor, attention_mask: Tensor,
+                token_type_ids: Tensor, position_ids: Tensor) -> Tensor:
+        """Wrap the input and output to conform to the native process interface."""
+
+        return self._model(**{
+            'input_ids': input_ids,
+            'attention_mask': attention_mask,
+            'token_type_ids': token_type_ids,
+        })
+
+
+class HuggingFaceTransformerModel:
+
+    class TraceableModel(ABC):
+        def __init__(
+                self,
+                tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast],
+                model: Union[PreTrainedModel, SentenceTransformerWrapper, DPREncoderWrapper, DistilBertWrapper]
+        ):
+            self._tokenizer = tokenizer
+            self._model = model
+
+        def classification_labels(self) -> Optional[List[str]]:
+            return None
+
+        @abstractmethod
+        def trace(self) -> TracedModelTypes:
+            ...
+
+    class TraceableClassificationModel(TraceableModel, ABC):
+        def classification_labels(self) -> Optional[List[str]]:
+            labels = HuggingFaceTransformerModel._dict_to_ordered_list(self._model.config.id2label, sort_by_key=True)
+            # Make classes like I-PER into I_PER which fits Java enumerations
+            return [x.replace('-', '_') for x in labels]
+
+    class FillMaskModel(TraceableModel):
+        def trace(self) -> TracedModelTypes:
+            # model needs to be in evaluate mode
+            self._model.eval()
+
+            # tokenizing dummy input text
+            text = "[CLS] Who was Jim Henson ? [SEP] [MASK] Henson was a puppeteer [SEP]"
```

Comment:
This looks like real data, are these supposed to be parameters?

Reply:
Nope 😄  This does need to be improved but yes, you have to have some hard-coded examples that are used.